mediorum/server/serve_legacy.go

```diff
@@ -167,35 +167,77 @@ func (ss *MediorumServer) redirectToCid(c echo.Context, cid string, checkAllNode
-	// 			continue
-	// 		}
-	// 		req.Header.Set("User-Agent", "mediorum "+ss.Config.Self.Host)
-	// 		resp, err := http.DefaultClient.Do(req)
-	// 		if err != nil {
-	// 			logger.Error("error sending HEAD request", "err", err)
-	// 			continue
-	// 		}
-	// 		if resp.StatusCode == 200 || resp.StatusCode == 204 || resp.StatusCode == 206 || resp.StatusCode == 302 || resp.StatusCode == 304 {
-	// 			dest.Query().Add("localOnly", "true")
-	// 			return c.Redirect(302, dest.String())
-	// 		}
-	// 	}
-	// }
+	// don't check additional nodes beyond what's in cid_lookup if "localOnly" is true
+	if c.QueryParam("localOnly") == "true" {
+		return c.String(404, "not redirecting because localOnly=true")
+	}
+
+	// check random healthy hosts via HEAD request to see if they have the cid but aren't in our cid_lookup
+	if numRandNodesToCheck > len(healthyHosts) {
+		numRandNodesToCheck = len(healthyHosts)
+	}
+	source := rand.NewSource(time.Now().UnixNano())
+	r := rand.New(source)
+	r.Shuffle(len(healthyHosts), func(i, j int) { healthyHosts[i], healthyHosts[j] = healthyHosts[j], healthyHosts[i] })
+
+	randHosts := healthyHosts[:numRandNodesToCheck]
+	for _, host := range randHosts {
+		if host == ss.Config.Self.Host {
+			continue
+		}
+		dest := ss.replaceHost(c, host)
+		dest.Query().Add("localOnly", "true")
+
+		req, err := http.NewRequest("HEAD", dest.String(), nil) // NOTE: cloudflare seems to turn most of these HEADs into GETs
+		if err != nil {
+			logger.Error("error creating HEAD request", "err", err)
+			continue
+		}
+		req.Header.Set("User-Agent", "mediorum "+ss.Config.Self.Host)
+
+		client := &http.Client{
+			// without this option, we'll incorrectly think Node A has it when really Node A was telling us to redirect to Node B
+			CheckRedirect: func(req *http.Request, via []*http.Request) error {
+				return http.ErrUseLastResponse
+			},
+			Timeout: 10 * time.Second,
+		}
+		resp, err := client.Do(req)
+		if err != nil {
+			logger.Error("error sending HEAD request", "err", err)
+			continue
+		}
+
+		// if we found a healthy host with the cid, add it to our cid_lookup table and redirect
+		if resp.StatusCode == 200 || resp.StatusCode == 204 || resp.StatusCode == 206 {
+			ss.pgPool.Exec(ctx, `insert into cid_lookup ("multihash", "host") values ($1, $2) on conflict do nothing`, cid, host)
+			return c.Redirect(302, dest.String())
+		}
+
+		// if we found a host that knows where the cid is, follow their redirect and add it to our cid_lookup table
+		if resp.StatusCode == 302 || resp.StatusCode == 304 {
+			redirectLocation := resp.Header.Get("Location")
+			if redirectLocation == "" {
+				continue
+			}
+			redirectUrl, err := url.Parse(redirectLocation)
+			if err != nil || redirectUrl.Host == "" {
+				logger.Error("error parsing redirectLocation, or empty host", "err", err, "redirectLocation", redirectLocation)
+				continue
+			}
+
+			redirectHost := ss.getScheme() + "://" + redirectUrl.Host
```

Comment:
also seems like this could be `redirectUrl.Scheme + "://" + redirectUrl.Host`?

Reply:
yeah I think that would also work. from what I understand we're always https on stage and prod